Source/Parsing/Frame/Content/Operation/FrameContentWithSubframesParsingOperation.swift

```diff
@@ -0,0 +1,12 @@
+//
+//  FrameContentWithSubframesParsingOperation.swift
+//
+//  Created by Fabrizio Duroni on 20/02/2018.
+//  2018 Fabrizio Duroni.
+//
+
+import Foundation
+
+protocol FrameContentWithSubframesParsingOperation {
```

Comment:
I have a small quandary. Per your suggestion, I created a `subframePseudoTag` from the remaining `frameData` in `FrameParser.extractChapterElements` and created a variation on `ID3TagParser` for it.

The problem is that if I add `ID3SubframePseudoTagParser` (idk if I should come up with a better name for that?) into the parameters of the `extractChapterElements` method and then into the parameters of `ID3ChapterFrameContentParsingOperation.parse`, then `ID3ChapterFrameContentParsingOperation` no longer conforms with the `FrameContentParsingOperation` protocol.

To get around that, I created a `FrameContentWithSubframesParsingOperation` protocol, however, it turns out that the return for `ID3FrameContentParsingOperationFactory` is `[FrameType : FrameContentParsingOperation]`, so either I need `ID3FrameContentParsingOperationFactory` to return both `[FrameType : FrameContentParsingOperation]` AND `[FrameType : FrameContentWithSubframesParsingOperation]`, or I need another way around the compliance issue with `FrameContentParsingOperation`.

I mean, I could add an optional subframe handling parameter to `FrameContentParsingOperation` but that would require going in to each and every `ParsingOperationFactory` for each frame type and adding the parameter there as well, right?

Reply:
I guess it’s not as easy as it sounded at first.

> then ID3ChapterFrameContentParsingOperation no longer conforms with the FrameContentParsingOperation protocol.

Maybe the answer is to change the protocol requirement? And then provide a default for the sake of all the other types that already conform? But I’m admittedly confused by all the similarly named types.

Reply:
Since subframes are optional even in the CHAP and CTOC frames, can I get away with this?

```swift
protocol FrameContentParsingOperation {
    func parse(
        frame: Data,
        version: ID3Version,
        subframePseudoTagParser: ID3SubframePseudoTagParser?,
        completed: (FrameName, ID3Frame) -> ())
}
```

And then passing in `nil` for the frames that don't use it, like here?

```swift
class ID3RecordingTimeFrameContentParsingOperationTest: XCTestCase {
    func testframeContentValidTimestamp() {
        let expectation = XCTestExpectation(description: "recording time")
        let recordingTimeParsingOperation = ID3RecordingTimeFrameContentParsingOperation(
            stringContentParser: ID3FrameStringContentParserFactory.make()
        )
        
        recordingTimeParsingOperation.parse(
            frame: frameV4utf8Valid(),
            version: .version4,
            subframePseudoTagParser: nil,
            completed: {(frameName, frame) in
                XCTAssertEqual(frameName, .RecordingDateTime)
                XCTAssertEqual((frame as? ID3FrameRecordingDateTime)?.recordingDateTime.date?.day, 27)
                XCTAssertEqual((frame as? ID3FrameRecordingDateTime)?.recordingDateTime.date?.month, 7)
                XCTAssertEqual((frame as? ID3FrameRecordingDateTime)?.recordingDateTime.date?.year, 2018)
                XCTAssertEqual((frame as? ID3FrameRecordingDateTime)?.recordingDateTime.time?.hour, 11)
                XCTAssertEqual((frame as? ID3FrameRecordingDateTime)?.recordingDateTime.time?.minute, 35)
                expectation.fulfill()
                
        })
    }
```

Or am I going to turn around and find I've broken everything if I commit that?

Reply:
Only one way to find out.

Reply:
Well, it builds, and that particular test passes, so...  : fingerscrossed : (and do they seriously not have a crossed fingers emoji. Here, of all places?)

Reply:
U+1F91E: 🤞

Reply:
Alas, the crossed fingers were for naught...